realm/realm-library/src/main/cpp/io_realm_internal_OsSet.cpp

```diff
@@ -188,6 +205,208 @@ Java_io_realm_internal_OsSet_nativeRemoveString(JNIEnv* env, jclass, jlong set_p
     return nullptr;
 }
 
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_OsSet_nativeIsSubSetOf(JNIEnv* env, jclass, jlong set_ptr,
+                                              jlong other_set_ptr) {
+    try {
+        auto &other_set = *reinterpret_cast<realm::object_store::Set *>(other_set_ptr);
+        auto &set = *reinterpret_cast<realm::object_store::Set *>(set_ptr);
+        return other_set.is_subset_of(set);
+    }
+    CATCH_STD()
+    return false;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_OsSet_nativeContainsAllString(JNIEnv* env, jclass, jlong set_ptr,
+                                                     jobjectArray j_other_set_values) {
+    try {
+        auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+        JObjectArrayAccessor<JStringAccessor, jstring> values(env, j_other_set_values);
+        jsize values_size = values.size();
+        for (int i = 0; i < values_size; i++) {
+            JStringAccessor value = values[i];
+            size_t found;
+            if (value.is_null()) {
+                found = set.find_any(Mixed());
+            } else {
+                found = set.find_any(Mixed(StringData(value)));
+            }
+            if (found == npos) {    // npos represents "not found"
+                return false;
+            }
+        }
+        return true;
+    }
+    CATCH_STD()
+    return false;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_OsSet_nativeUnion(JNIEnv*, jclass, jlong set_ptr, jlong other_set_ptr) {
+    auto& set = *reinterpret_cast<realm::object_store::Set*>(set_ptr);
+    auto& other_set = *reinterpret_cast<realm::object_store::Set*>(other_set_ptr);
+
+    // If other set is a subset of set it means set will not change after the union
+    bool has_changed = !other_set.is_subset_of(set);
```

Comment:
Sound like some information that could just be returned from the below call instead. Maybe collect a list of all these nice to have optimizations and pass them to core.